jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/Invocable.java

```diff
@@ -51,24 +53,62 @@ enum InvocationType
          * <p>This invocation type is suitable for {@code Invocable}s that
          * call application code, for example to process an HTTP request.</p>
          */
-        BLOCKING,
+        BLOCKING
+        {
+            public void runWithoutBlocking(Runnable task, Executor executor)
+                {
+                    executor.execute(task);
+                }
+        },
         /**
          * <p>Invoking the {@link Invocable} does not block the invoker thread,
-         * and the invocation may be performed immediately in the invoker thread.</p>
+         * and the invocation must be performed immediately in the invoker thread.</p>
          * <p>This invocation type is suitable for {@code Invocable}s that
-         * call implementation code that is guaranteed to never block the
+         * call high priority implementation code that is guaranteed to never block the
          * invoker thread.</p>
          */
-        NON_BLOCKING,
+        NON_BLOCKING
+        {
+            public void runWithoutBlocking(Runnable task, Executor ignored)
+                {
+                    task.run();
+                }
+        },
         /**
-         * <p>Invoking the {@link Invocable} may block the invoker thread,
-         * but the invocation cannot be deferred to a later time, differently
-         * from {@link #BLOCKING}.</p>
-         * <p>This invocation type is suitable for {@code Invocable}s that
-         * themselves perform the non-deferrable action in a non-blocking way,
-         * thus advancing a possibly stalled system.</p>
+         * <p>Invoking the {@link Invocable} may act as a {@code BLOCKING} invocable, unless
+         * it is called via {@link Invocable#invokeNonBlocking(Runnable)}, in which case it must
+         * act as a {@code NON_BLOCKING}.  The implementation must check {@link Invocable#isNonBlockingInvocation()}
+         * to determine how it was called.</p>
+         * <p>This invocation type is suitable for {@code Invocable}s that have high priority actions that cannot be deferred,
+         * even if the caller is constrained in a way that it cannot immediately call a {@code BLOCKING} invocable, but that
+         * may also have lower priority blocking actions that can benefit from being directly called.
```

Comment:
I don't like "high" and "low" priority, they seem arbitrary and not well defined.

I think the main reason for which EITHER tasks cannot be deferred is that they produce (or cause the production of) subtasks, and they must do so otherwise the system is idle/hung. This was present in my version.